### PR TITLE
feat(FN-3377): fix report received email template TFM URL

### PR DESCRIPTION
**Pull request by sw-mattw**
## Introduction :pencil2:
At present, the TFM URL in the `UKEF Report Received` template is hardcoded to the production deployment. We wish for this to update automatically based on the environment (i.e. use the feature deployment URL for feature, dev URL for dev, etc).

## Resolution :heavy_check_mark:
- Additional parameter/"placeholder" added to the email template to allow the URL value to be passed in. This value is taken from the `TFM_UI_URL` environment variable, the same way task URLs are generated in TFM API (see referenced files below).
- Docker compose and deployment jobs updated to include the new environment variable for Portal API (in the same way that it is included for use in TFM API).

Changes outside PR:
- New template added to GOV.UK Notify which accepts the new "placeholder" for the URL. Old template kept separate to ensure backwards compatability.

Files relating to task URL generation in TFM API (referenced above):
- `trade-finance-manager-api\src\v1\controllers\send-deal-submit-emails.js:34`
- `trade-finance-manager-api\src\v1\helpers\amendment.helpers.js:323`

Resulting email when tested locally, by uploading a valid report to portal, with `TFM_UI_URL=http://localhost:5003`:

Screenshot 2024-09-04 173900

 
## Miscellaneous :heavy_plus_sign:
N/A

### CODE REVIEW

File: .github/workflows/development_infrastructure.yml
Comment (sw-mattw) [inline .github/workflows/development_infrastructure.yml:1]: @ChristianMcc @abhi-markan would you mind reviewing these infrastructure and docker changes please to see if they look correct, thanks

File: portal-api/src/v1/services/utilisation-report/email-service.ts [hunk 1]
```diff
@@ -17,17 +17,35 @@ const getUkefGefReportingEmailRecipients = () => {
   }
 };
 
+/**
+ * Retrieves the TFM UI URL from the environment variables.
+ * @returns The TFM UI URL.
+ * @throws {InvalidEnvironmentVariableError} If the TFM_UI_URL environment variable is not defined or is empty.
+ */
+const getTfmUiUrl = () => {
```
Comment (JoshBinns2000) [inline portal-api/src/v1/services/utilisation-report/email-service.ts:25]: Could we explicitly annotate the return type to `string` here?

File: portal-api/src/v1/services/utilisation-report/email-service.ts [hunk 2]
```diff
@@ -17,17 +17,35 @@ const getUkefGefReportingEmailRecipients = () => {
...
+  }
+
+  return tfmUiUrl;
+};
+
 /**
  * Sends notification email to UKEF GEF reporting email recipients that a utilisation report has been submitted
  * @param bankName - name of the bank
  * @param reportPeriod - period for which the report covers as a string, eg. June 2023
+ * @param tfmHomepageUrl - URL for the TFM homepage
```
Comment (JoshBinns2000) [inline portal-api/src/v1/services/utilisation-report/email-service.ts:38]: I don't think this is a param anymore :)

File: portal-api/src/v1/services/utilisation-report/email-service.ts (hunk 1 above)
Comment (JoshBinns2000) [inline portal-api/src/v1/services/utilisation-report/email-service.ts:25]: Could we also add some unit tests for this function?

File: portal-api/src/v1/services/utilisation-report/email-service.test.ts [hunk 3]
```diff
@@ -53,12 +57,23 @@ describe('emailService', () => {
...
       // Arrange
+      process.env.TFM_UI_URL = 'https://www.ukexportfinance.gov.uk';
       delete process.env.UKEF_GEF_REPORTING_EMAIL_RECIPIENT;
 
       // Act + Assert
       await expect(sendUtilisationReportUploadNotificationEmailToUkefGefReportingTeam('My Bank', 'June 2026')).rejects.toThrow();
       expect(sendEmail).toHaveBeenCalledTimes(0);
     });
+
+    it('should throw an error if TFM_UI_URL is undefined', async () => {
```
Comment (JoshBinns2000) [inline portal-api/src/v1/services/utilisation-report/email-service.test.ts:68]: I think this should be tested as part of a unit test for `getTfmUiUrl` and then we should mock that function in this unit test

File: portal-api/src/v1/services/utilisation-report/email-service.test.ts [hunk 4]
```diff
@@ -53,12 +57,23 @@ describe('emailService', () => {
...
+      // Arrange
+      delete process.env.TFM_UI_URL;
+      process.env.UKEF_GEF_REPORTING_EMAIL_RECIPIENT = '["email1@ukexportfinance.gov.uk"]';
+
+      // Act + Assert
+      await expect(sendUtilisationReportUploadNotificationEmailToUkefGefReportingTeam('My Bank', 'June 2026')).rejects.toThrow();
```
Comment (JoshBinns2000) [inline portal-api/src/v1/services/utilisation-report/email-service.test.ts:74]: Could we explicitly assert that the thrown error is the `InvalidEnvironmentVariableError` error?

**Review by JoshBinns2000 — COMMENTED** [review]
Just a few comments, but nice solution

File: portal-api/src/v1/services/utilisation-report/email-service.ts (hunk 2 above)
Comment (sw-mattw) [reply portal-api/src/v1/services/utilisation-report/email-service.ts]: Thanks for catching, forgot to remove this after refactoring

File: portal-api/src/v1/services/utilisation-report/email-service.test.ts (hunk 3 above)
Comment (sw-mattw) [reply portal-api/src/v1/services/utilisation-report/email-service.test.ts]: Updated this function to use a spy thanks

File: portal-api/src/v1/services/utilisation-report/email-service.test.ts [hunk 5]
```diff
@@ -65,14 +69,21 @@ describe('emailService', () => {
...
-    it('should throw an error if TFM_UI_URL is undefined', async () => {
+    it('should throw an InvalidEnvironmentVariable error if TFM_UI_URL is undefined', async () => {
       // Arrange
-      delete process.env.TFM_UI_URL;
       process.env.UKEF_GEF_REPORTING_EMAIL_RECIPIENT = '["email1@ukexportfinance.gov.uk"]';
 
+      const mockGetTfmUiUrl = jest.fn().mockImplementation(() => {
+        throw new InvalidEnvironmentVariableError('Invalid TFM_UI_URL');
+      });
+      const getTfmUiUrlSpy = jest.spyOn(emailService, 'getTfmUiUrl').mockImplementation(mockGetTfmUiUrl);
```
Comment (JoshBinns2000) [inline portal-api/src/v1/services/utilisation-report/email-service.test.ts:79]: You should be able to write this as

```typescript
const getTfmUiUrlSpy = jest.spyOn(emailService, 'getTfmUiUrl').mockImplementation(() => {
    throw new InvalidEnvironmentVariableError('Invalid TFM_UI_URL');
});
...
expect(getTfmUiUrlSpy).toHaveBeenCalledTimes(1);
```

File: portal-api/src/v1/services/utilisation-report/email-service.test.ts [hunk 6]
```diff
@@ -65,14 +69,21 @@ describe('emailService', () => {
...
+
       // Act + Assert
-      await expect(sendUtilisationReportUploadNotificationEmailToUkefGefReportingTeam('My Bank', 'June 2026')).rejects.toThrow();
+      await expect(sendUtilisationReportUploadNotificationEmailToUkefGefReportingTeam('My Bank', 'June 2026')).rejects.toThrow(InvalidEnvironmentVariableError);
+      expect(mockGetTfmUiUrl).toHaveBeenCalledTimes(1);
       expect(sendEmail).toHaveBeenCalledTimes(0);
+
+      getTfmUiUrlSpy.mockClear();
```
Comment (JoshBinns2000) [inline portal-api/src/v1/services/utilisation-report/email-service.test.ts:86]: I don't think this should be needed

File: portal-api/src/v1/services/utilisation-report/email-service.test.ts [hunk 7]
```diff
@@ -1,8 +1,11 @@
+import { InvalidEnvironmentVariableError } from '@ukef/dtfs2-common';
 import { BankResponse } from '../../api-response-types';
 import {
+  getTfmUiUrl,
   sendUtilisationReportUploadConfirmationEmailToBankPaymentOfficerTeam,
   sendUtilisationReportUploadNotificationEmailToUkefGefReportingTeam,
 } from './email-service';
+import * as emailService from './email-service';
```
Comment (JoshBinns2000) [inline portal-api/src/v1/services/utilisation-report/email-service.test.ts:8]: I would be tempted to write this as
```typescript
import * as emailService from './email-service';

...

const {
  getTfmUiUrl,
  sendUtilisationReportUploadConfirmationEmailToBankPaymentOfficerTeam,
  sendUtilisationReportUploadNotificationEmailToUkefGefReportingTeam,
} = emailService;
```

**Review by JoshBinns2000 — COMMENTED** [review]
Just a few more comments - sorry

**Review by JoshBinns2000 — APPROVED** [review]
LGTM

File: portal-api/src/v1/services/utilisation-report/email-service.ts [hunk 8]
```diff
@@ -17,17 +17,34 @@ const getUkefGefReportingEmailRecipients = () => {
   }
 };
 
+/**
+ * Retrieves the TFM UI URL from the environment variables.
+ * @returns The TFM UI URL.
+ * @throws {InvalidEnvironmentVariableError} If the TFM_UI_URL environment variable is not defined or is empty.
+ */
+export const getTfmUiUrl = (): string => {
+  const tfmUiUrl = process.env.TFM_UI_URL;
```
Comment (abhi-markan) [inline portal-api/src/v1/services/utilisation-report/email-service.ts:26]:
```suggestion
  const { TFM_UI_URL } = process.env;
```

**Review by abhi-markan — CHANGES_REQUESTED** [review]
Looks good, one minor comment. Thanks

File: portal-api/src/v1/services/utilisation-report/email-service.ts (hunk 8 above)
Comment (sw-mattw) [reply portal-api/src/v1/services/utilisation-report/email-service.ts]: Updated thank you

**Review by abhi-markan — APPROVED** [review]
LGTM 👍🏼

File: portal-api/src/v1/services/utilisation-report/email-service.ts (hunk 8 above)
Comment (sw-mattw) [reply portal-api/src/v1/services/utilisation-report/email-service.ts]: _PR approved following this change, this was the only comment, resolving so I can merge_

**Pull request merged**